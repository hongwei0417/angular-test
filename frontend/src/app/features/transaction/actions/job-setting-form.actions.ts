import { createAction, props } from '@ngrx/store';

export const addJobSetting = createAction(
  '[JobSetting] add a job setting option'
);

export const removeJobSetting = createAction(
  '[JobSetting] remove a job setting option',
  props<{ id: number }>()
);
