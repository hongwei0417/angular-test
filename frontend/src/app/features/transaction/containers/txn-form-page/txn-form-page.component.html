<div class="container">
  <div class="title">
    <ng-container [ngSwitch]="formType$ | async">
      <h2 *ngSwitchCase="txnFormType.CREATE">Create Transaction</h2>
      <h2 *ngSwitchCase="txnFormType.VIEW">View Transaction</h2>
      <h2 *ngSwitchCase="txnFormType.EDIT">Edit Transaction</h2>
      <h2 *ngSwitchCase="txnFormType.COPY">Copy Transaction</h2>
    </ng-container>
    <button class="txtBtn customBtn arrow" routerLink="/transaction/list">
      Back
    </button>
  </div>
  <app-txn-setting></app-txn-setting>
  <app-job-setting></app-job-setting>
</div>
