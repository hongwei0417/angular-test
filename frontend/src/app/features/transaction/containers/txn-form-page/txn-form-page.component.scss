@import "/src/assets/styles/variables.scss";

.title {
  border-bottom: 1px solid $gray-500;
  padding-bottom: 10px;
}

h3 {
  display: flex;
  align-items: center;
  text-align: center;
  color: $primary-medium;
  margin-bottom: 10px;
}


h3:before {
  content: "1";
  margin-right: 5px;
  width: 25px;
  height: 25px;
  display: block;
  border-radius: 50%;
  background: $primary-medium;
  color: $gray-100;
  line-height: 25px;
  font-size: 16px;
}


input {
  background: $gray-100;
}


.btnGroup {
  display: flex;
}

.btnGroup button:last-child {
  margin-left: 15px;
}


