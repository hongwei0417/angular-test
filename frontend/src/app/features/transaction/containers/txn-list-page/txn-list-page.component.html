<h3 class="text-center">Transaction List Page</h3>

<div class="d-flex justify-content-center">
  <button [routerLink]="['/transaction', 'list']">Transaction list</button>
  <button [routerLink]="['/transaction', 'create']">Create Transaction</button>
</div>

<div class="d-flex justify-content-center">
  <button id="generateDataBtn" (click)="loadTxnData()">Generate Data</button>
</div>

<app-txn-list-table
  [tableData]="[]"
  (removeEvent)="onRemoveTxn($event)"
></app-txn-list-table>
