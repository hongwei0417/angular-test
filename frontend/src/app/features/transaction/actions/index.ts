import * as TxnApiActions from './txn-api.actions';
import * as TxnFormPageActions from './txn-form-page.actions';
import * as TxnListPageActions from './txn-list-page.actions';
import * as TxnSettingFormActions from './txn-setting-form.actions';

export {
  TxnApiActions,
  TxnFormPageActions,
  TxnListPageActions,
  TxnSettingFormActions,
};
