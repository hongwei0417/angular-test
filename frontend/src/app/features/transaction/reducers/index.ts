import {
  Action,
  ActionReducer,
  ActionReducerMap,
  combineReducers,
  createFeatureSelector,
  createSelector,
  MetaReducer,
} from '@ngrx/store';
import * as fromRoot from '../../../core/reducers';
import * as fromTxnListPage from './txn-list-page.reducer';
import * as fromTxnFormPage from './txn-form-page.reducer';
import * as fromTxnSettingForm from './txn-setting-form.reducer';
import * as fromJobSettingForm from './job-setting-form.reducer';

export const FeatureKey = 'transaction';

export interface TxnState {
  [fromTxnListPage.FeatureKey]: fromTxnListPage.State;
  [fromTxnFormPage.FeatureKey]: fromTxnFormPage.State;
  txnFormState: TxnFormState;
}
export interface TxnFormState {
  [fromTxnSettingForm.FeatureKey]: fromTxnSettingForm.State;
  [fromJobSettingForm.FeatureKey]: fromJobSettingForm.State;
}
export interface State extends fromRoot.State {
  [FeatureKey]: TxnState;
}

export const txnFormReducers = (
  state: TxnFormState | undefined,
  action: Action
) =>
  combineReducers({
    [fromTxnSettingForm.FeatureKey]: fromTxnSettingForm.reducer,
    [fromJobSettingForm.FeatureKey]: fromJobSettingForm.reducer,
  })(state, action);

/** Provide reducer in AoT-compilation happy way */
export const reducers = (state: TxnState | undefined, action: Action) => {
  return combineReducers({
    [fromTxnListPage.FeatureKey]: fromTxnListPage.reducer,
    [fromTxnFormPage.FeatureKey]: fromTxnFormPage.reducer,
    txnFormState: txnFormReducers,
  })(state, action);
};

// feature selectors
export const getTxnState = createFeatureSelector<State, TxnState>(FeatureKey);

// sub reducers basic state
export const getTxnListPageState = createSelector(
  getTxnState,
  (state: TxnState) => state[fromTxnListPage.FeatureKey]
);

export const getTxnFormPageState = createSelector(
  getTxnState,
  (state: TxnState) => state[fromTxnFormPage.FeatureKey]
);

export const getTxnSettingFormState = createSelector(
  getTxnState,
  (state: TxnState) => state.txnFormState[fromTxnSettingForm.FeatureKey]
);

export const getJobSettingFormState = createSelector(
  getTxnState,
  (state: TxnState) => state.txnFormState[fromJobSettingForm.FeatureKey]
);

// selectors
export const {
  selectAll: getAllTxns,
  selectEntities: getTxnEntities,
  selectIds: getTxnIds,
  selectTotal: getTotalTxns,
} = fromTxnListPage.adapter.getSelectors(getTxnListPageState);

export const getLatestTxnId = createSelector(
  getTxnListPageState,
  fromTxnListPage.getLatestId
);

export const getLatestTxn = createSelector(
  getTxnEntities,
  getLatestTxnId,
  (entities, latestId) => {
    return latestId && entities[latestId];
  }
);

export const getFilterForm = createSelector(
  getTxnListPageState,
  fromTxnListPage.getFilterForm
);

export const getTxnSettingForm = createSelector(
  getTxnSettingFormState,
  fromTxnSettingForm.getFormState
);

export const getFqSettingOptions = createSelector(
  getTxnSettingFormState,
  fromTxnSettingForm.getFqOptions
);

export const getFqShowAccordions = createSelector(
  getTxnSettingFormState,
  fromTxnSettingForm.getShowAccordions
);

export const getModuleOptions = createSelector(
  getTxnSettingFormState,
  fromTxnSettingForm.getModuleOptions
);

export const getJobSettingForm = createSelector(
  getJobSettingFormState,
  fromJobSettingForm.getFormState
);

export const getSingleJobOptions = createSelector(
  getJobSettingFormState,
  fromJobSettingForm.getSingleJobOptions
);
