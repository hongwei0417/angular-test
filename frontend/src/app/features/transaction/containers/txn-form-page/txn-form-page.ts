import { Observable, of } from 'rxjs';
import { Store, select } from '@ngrx/store';
import { Component, OnDestroy, OnInit } from '@angular/core';
import { TxnFormPageActions } from '../../actions';
import * as fromTxn from '../../reducers';

@Component({
  selector: 'app-txn-form-page',
  templateUrl: './txn-form-page.component.html',
  styleUrls: ['./txn-form-page.component.scss'],
})
export class TxnCreatePageComponent implements OnInit, OnDestroy {
  error$!: Observable<string>;

  constructor(private store$: Store<fromTxn.State>) {}

  ngOnInit(): void {}

  ngOnDestroy(): void {
    this.store$.dispatch(TxnFormPageActions.clearTxnCreatePageState());
  }

  onSubmit({ title, content, executeCount }: any): void {
    this.store$.dispatch(
      TxnFormPageActions.createTxn({ title, content, executeCount })
    );
  }
}
