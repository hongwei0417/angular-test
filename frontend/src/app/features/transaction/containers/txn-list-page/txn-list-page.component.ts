import { FormGroupState } from 'ngrx-forms';
import { FilterFormValue } from './../../reducers/txn-list-page.reducer';
import { Observable, of } from 'rxjs';
import { Component, OnInit, OnDestroy } from '@angular/core';
import { select, Store } from '@ngrx/store';
import * as fromTxn from '../../reducers';
import { TxnListPageActions } from '../../actions';
import { TxnListTableRow } from '../../models/TxnListTable';
@Component({
  selector: 'app-txn-list-page',
  templateUrl: './txn-list-page.component.html',
  styleUrls: ['./txn-list-page.component.scss'],
})
export class TxnListPageComponent implements OnInit, OnDestroy {
  txnData$!: Observable<TxnListTableRow[]>;
  filterForm$!: Observable<FormGroupState<FilterFormValue>>;

  constructor(private store$: Store<fromTxn.State>) {}

  ngOnInit(): void {
    this.txnData$ = this.store$.pipe(select(fromTxn.getAllTxns));
    this.filterForm$ = this.store$.pipe(select(fromTxn.getFilterForm));
    this.store$.dispatch(TxnListPageActions.loadTxnList());
  }

  ngOnDestroy(): void {
    this.store$.dispatch(TxnListPageActions.clearTxnListPageState());
  }

  createNewTxn(): void {
    this.store$.dispatch(TxnListPageActions.createNewTxn());
  }

  onViewTxnInfo(id: string): void {
    this.store$.dispatch(TxnListPageActions.viewTxnInfo({ id }));
  }

  onEditTxnInfo(id: string): void {
    this.store$.dispatch(TxnListPageActions.editTxnInfo({ id }));
  }

  onCopyInfo(id: string): void {
    this.store$.dispatch(TxnListPageActions.copyTxnInfo({ id }));
  }

  onExecuteTxn(id: string): void {
    this.store$.dispatch(TxnListPageActions.executeTxn({ id }));
  }

  onReloadTxn(id: string): void {
    this.store$.dispatch(TxnListPageActions.reloadTxn({ id }));
  }
}
