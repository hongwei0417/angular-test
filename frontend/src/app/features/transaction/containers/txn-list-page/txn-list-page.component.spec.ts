import { RouterLinkDirectiveStub } from './../../../../shared/testing/router-link-directive-stub.directive';
import { TxnCreatePageComponent } from '../txn-form-page/txn-form-page';
import { fakeData } from '../../../../core/services/fake-data.service';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { DebugElement, NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import * as fromTxn from '../../reducers';

import { TxnListPageComponent } from './txn-list-page.component';
import { createSelector } from '@ngrx/store';
import { TxnListPageActions } from '../../actions';
import { By } from '@angular/platform-browser';

describe('TransactionListComponent', () => {
  let fixture: ComponentFixture<TxnListPageComponent>;
  let listPageElement: HTMLElement;
  let component: TxnListPageComponent;
  let store$: MockStore;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [],
      declarations: [
        TxnListPageComponent,
        TxnCreatePageComponent,
        RouterLinkDirectiveStub,
      ],
      providers: [
        provideMockStore({
          selectors: [
            {
              selector: createSelector(fromTxn.getAllTxns, (state) => state),
              value: fakeData,
            },
          ],
        }),
      ],
      schemas: [NO_ERRORS_SCHEMA],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(TxnListPageComponent);
    component = fixture.componentInstance;
    listPageElement = fixture.nativeElement;
    store$ = TestBed.inject(MockStore);
    fixture.detectChanges();
    jest.spyOn(store$, 'dispatch');
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  // it('should dispatch a TxnListPage.LoadTxnList action on click', () => {
  //   const action = TxnListPageActions.loadTxnList();
  //   const btnElement = listPageElement.querySelector(
  //     '#generateDataBtn'
  //   ) as HTMLElement;
  //   btnElement.click();
  //   expect(store$.dispatch).toHaveBeenCalledWith(action);
  // });

  // it('should navigate to create txn page on click', () => {
  //   const linkDes = fixture.debugElement.queryAll(
  //     By.directive(RouterLinkDirectiveStub)
  //   );
  //   const routerLinks = linkDes.map((de) =>
  //     de.injector.get(RouterLinkDirectiveStub)
  //   );
  //   linkDes[1].triggerEventHandler('click', null);
  //   fixture.detectChanges();
  //   expect(routerLinks[1].navigatedTo).toEqual(['/transaction', 'create']);
  // });
});
