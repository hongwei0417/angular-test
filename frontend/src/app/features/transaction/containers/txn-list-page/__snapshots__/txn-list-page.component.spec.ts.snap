// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionListComponent should create 1`] = `
<app-txn-list-page
  store$={[Function MockStore]}
  txnData$={[Function Store]}
>
  <h3
    class="text-center"
  >
    Transaction List Page
  </h3><div
    class="d-flex justify-content-center"
  >
    <button>
      Transaction list
    </button>
    <button>
      Create Transaction
    </button>
  </div><div
    class="d-flex justify-content-center"
  >
    <button
      id="generateDataBtn"
    >
      Generate Data
    </button>
  </div><app-txn-list-table />
</app-txn-list-page>
`;

exports[`TransactionListComponent should match snapshot 1`] = `
<app-txn-list-page
  store$={[Function MockStore]}
  txnData$={[Function Store]}
>
  <h3
    class="text-center"
  >
    Transaction List Page
  </h3><div
    class="d-flex justify-content-center"
  >
    <button>
      Transaction list
    </button>
    <button>
      Create Transaction
    </button>
  </div><div
    class="d-flex justify-content-center"
  >
    <button
      id="generateDataBtn"
    >
      Generate Data
    </button>
  </div><app-txn-list-table />
</app-txn-list-page>
`;
